### PR TITLE
Add functionality in TextCorpus to convert document text to index vectors

### PR DESCRIPTION
**Pull request by roopalgarg**
TextCorpus doesn't provide a way to convert document text to index vector as needed for say DL NLP models.
Adding a 'doc2idx' to Dictionary object and creating modes in TextCorpus to leverage this functionality.
Referencing issue #1634

### CODE REVIEW

**Comment by roopalgarg** [comment]
@menshikh-iv how does this look?

File: gensim/corpora/dictionary.py [hunk 1]
```diff
@@ -173,6 +173,37 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
         else:
             return result
 
+    def doc2idx(self, document, unk_wrd_idx=0):
```
[the remark this answers is not in the document]
Comment (menshikh-iv) [reply gensim/corpora/dictionary.py]: Also please rename `unk_wrd_idx` to `unknown_word_index` (here and everywhere)

File: gensim/corpora/textcorpus.py [hunk 2]
```diff
@@ -112,7 +114,10 @@ class TextCorpus(interfaces.CorpusABC):
     6.  remove stopwords; see `gensim.parsing.preprocessing` for the list of stopwords
 
     """
-    def __init__(self, input=None, dictionary=None, metadata=False, character_filters=None, tokenizer=None, token_filters=None):
+    def __init__(
```
Comment (menshikh-iv) [inline gensim/corpora/textcorpus.py]: Please use vertical indent (**only** for method/function definition, in all other cases - hanging indent).

File: gensim/corpora/dictionary.py [hunk 3]
```diff
@@ -173,6 +173,37 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
...
+
+        Replace all unknown words i.e, words not in the dictionary with the index as set via `unk_wrd_idx`, defaults to
+        0.
+
+        This function is `const`, aka read-only
+        """
+        if isinstance(document, string_types):
+            raise TypeError("doc2idx expects an array of unicode tokens on input, not a single string")
+
+        token2id = self.token2id
```
[the remark this answers is not in the document]
Comment (roopalgarg) [reply gensim/corpora/dictionary.py]: was just following convention from 'doc2bow' but I will make the change as you pointed out.

**Comment by roopalgarg** [comment]
@menshikh-iv the idea was that since we are adding the functionality to the `__iter__` of `TextCorpus`, we can process a corpus to get vocabulary indexes of the corpus. Also since `WikiCorpus` inherits `TextCorpus` and doesnt override the `__iter__` the functionality gets added to `WikiCorpus` as well.
Though I agree that the constructor for `WikiCorpus` doesnt explicitly take parameters for this new functionality, but we can set the parameters after creating an instance of the class by explicitly setting the new parameters.

So couple of things here then:

1. Should `WikiCorpus` have the parameters explicitly as well as part of its constructor definition?
2. Where else do you think this functionality should be added since you were asking as to why only `TextCorpus`?

**Comment by menshikh-iv** [comment]
@roopalgarg current problem is more "global", let me describe:

- We have many different classes for corpuses, we need to add it for all corpuses (but I don't think that this is a good idea, but have same interfaces will be nice).
- How to choose the concrete class for support this Dictionary feature (maybe we no need to add it to any corpuses, only new method for Dictionary?)
- Passing arguments (if we want to add it in many places)

I think corpus classes needs global refactoring (bring everything to the same interfaces and simplify, i.e. a minimum of functionality), @roopalgarg it isn't your problem, sorry, but you reminded me about the old and important problem.

@roopalgarg I'm ready to merge only new method for Dictionary right now.

@piskvorky wdyt? corpuses in common are very chaotic, have any idea how to rework it?

**Comment by roopalgarg** [comment]
@menshikh-iv I see your point. For now just adding `doc2idx` to the `Dictionary` class sounds good to me.
Do I need to revert my changes to the other 2 files (or there is a way you can discard the changes to them)?

File: gensim/corpora/dictionary.py [hunk 4]
```diff
@@ -173,6 +173,40 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
...
+        (either unicode or utf8-encoded).
+        No further preprocessing is done on the words in `document`; apply tokenization, stemming etc. before calling
+        this method.
+
+        Replace all unknown words i.e, words not in the dictionary with the index as set via `unknown_word_index`,
+        defaults to -1.
+
+        Notes
+        -----
+            This function is `const`, aka read-only
```
Comment (menshikh-iv) [inline gensim/corpora/dictionary.py]: No needed indentation here

File: gensim/corpora/dictionary.py [hunk 5]
```diff
@@ -173,6 +173,40 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
...
+
+        Parameters
+        ----------
+        document : list
```
Comment (menshikh-iv) [inline gensim/corpora/dictionary.py]: `list of str`

File: gensim/corpora/dictionary.py [hunk 6]
```diff
@@ -173,6 +173,40 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
...
+            List of words tokenized, normalized and preprocessed.
```
Comment (menshikh-iv) [inline gensim/corpora/dictionary.py]: No need to mention type twice

File: gensim/corpora/dictionary.py [hunk 7]
```diff
@@ -173,6 +173,40 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
...
+        unknown_word_index : int, optional
+            Index to use for words not in the dictionary.
+
+        Returns
+        -------
+        list
```
Comment (menshikh-iv) [inline gensim/corpora/dictionary.py]: `list of int`

File: gensim/corpora/dictionary.py [hunk 8]
```diff
@@ -173,6 +173,40 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
...
+            List of indexes in the dictionary for words in the `document`
```
Comment (menshikh-iv) [inline gensim/corpora/dictionary.py]: No need to mention type twice + add `preserves order`.

File: gensim/corpora/dictionary.py [hunk 9]
```diff
@@ -173,6 +173,40 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
...
+
```
Comment (menshikh-iv) [inline gensim/corpora/dictionary.py]: Please add example section (simple example that works how to apply this method)

**Comment by menshikh-iv** [comment]
@roopalgarg yeah, please revert 2 files, fix docstring and that's all :+1:

**Comment by roopalgarg** [comment]
@menshikh-iv a little new to numpy style docstrings so not fully aware of best practices. learnt something new today :)

**Comment by roopalgarg** [comment]
@menshikh-iv good to go ?

**Comment by roopalgarg** [comment]
@menshikh-iv good to merge ?

**Comment by menshikh-iv** [comment]
@roopalgarg yeah, thanks for your contribution:+1:

**Comment by roopalgarg** [comment]
@menshikh-iv awesome! thanks

**Pull request merged**